### PR TITLE
Add gandi provider

### PR DESCRIPTION
**Pull request by raphink**
Disclaimer: this is my first piece of code in golang, and I have no idea how to test this code properly.

Please let me know how I can ensure this works if this needs any adjustments!

### CODE REVIEW

**Comment by alena1108** [comment]
@raphink thank you for the contribution! There is a lack of tests on our side unfortunately, we are going to address it soon to make things easier for the contributors. 

Today the best way to test the provider - end to end by deploying the gandi as a Rancher service

1) build docker image locally from the build having your changes (example of the dockerfile for building the image can be found here:

https://github.com/rancher/external-dns/blob/master/packaging/Dockerfile

2) create docker-compose.yml template that would look similar to all dns providers templates under rancher-catalog (https://github.com/rancher/rancher-catalog/blob/master/templates/pointhq/1/docker-compose.yml). Example:

```
gandi:
  image: <your local image name>
  command: --provider gandi
  expose: 
   - 1000
  environment:
    GANDI_APIKEY: <api key>
    ROOT_DOMAIN: <root domain>
    TTL: 300
  labels:
    io.rancher.container.create_agent: "true"
    io.rancher.container.agent.role: "external-dns"
```


3) Deploy gandi stack in Rancher, verify that it comes up fine.
4) Add any other service with public port published to the host, and verify that gandi provider created a record for it. Scale up the service, verify that the gandi server got an extra ip address for the service record programmed. Remove this service, verify that the record is gone from gandi server.

If all the above works, this PR is good to go + we would have to submit a PR for rancher-catalog, so its enabled for Rancher users to deploy: https://github.com/rancher/rancher-catalog

**Comment by raphink** [comment]
@alena1108 thank you. I had started building the project and creating an image indeed, and am about to test it on my rancher server. I will let you know how this goes.

**Comment by raphink** [comment]
It seems I was not able to produce the tarball properly (I don't have `apply.sh` and some other files). I'm trying to just copy the binary from the Dockerfile, but it doesn't work great.

Also, how do you update the Godeps directory properly to include in the PR?

**Comment by alena1108** [comment]
@raphink 

>> how do you update the Godeps directory properly to include in the PR?

just run "godep save" under external-dns project, it should update the godeps. 


>> I'm trying to just copy the binary from the Dockerfile, but it doesn't work great.

Copying generated binary should be fine.  What exactly doesn't work? Most likely its due to missing godeps.

**Comment by raphink** [comment]
### Dependencies

I'm probably doing things wrong with go dependencies, since I'm new to this... I'm running `go get ./...` which works with my env being:

```bash
GOBIN=/home/rpinson/go/bin
GOROOT=/goroot
GOPATH=/home/rpinson/go
```

However, when I run `~/go/bin/godep save`, it says:

```bash
godep: WARNING: Recorded go version (go1.5.1) with minor version string found.
godep: WARNING: Recorded major go version (go1.5) and in-use major go version (go1.4) differ.
godep: To record current major go version run `godep update -goversion`.
godep: Package (golang.org/x/sys/unix) not found
```

### Dockerfile

As for the Dockerfile in `packaging/`, I've added an entrypoint script to debug what's happening:

```Dockerfile
FROM alpine:3.2
MAINTAINER Rancher Labs, Inc.
RUN apk add --update ca-certificates 

ENV EXT_DNS_RELEASE v0.2.1
ADD ./external-dns /usr/bin/external-dns
ADD ./entrypoint.sh /entrypoint.sh

ENTRYPOINT ["/entrypoint.sh"]
#ENTRYPOINT ["/usr/bin/external-dns"]
```

with `entrypoint.sh` begin:

```shell
#!/bin/sh

echo "Hello, world"

test -x /usr/bin/external-dns && echo "it's there"

ls -alh /usr/bin/external-dns
ldd /usr/bin/external-dns

exec /usr/bin/external-dns
```

Building and running gives me:

```shell
$ docker run -ti camptocamp/external-dns:latest 
Hello, world
it's there
-rwxrwxr-x    1 root     root       11.9M Jan 21 22:53 /usr/bin/external-dns
	/lib64/ld-linux-x86-64.so.2 (0x55ad1408b000)
	libpthread.so.0 => /lib64/ld-linux-x86-64.so.2 (0x55ad1408b000)
	libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x55ad1408b000)
/entrypoint.sh: exec: line 10: /usr/bin/external-dns: not found
```

and I can't figure what's wrong with that...

**Comment by raphink** [comment]
OK, when it comes ot the Dockerfile, the problem was that I was building a 64 bit executable, while the alpine image is 32 bit. Fixed by using `export GOARCH=386`.

**Comment by raphink** [comment]
@alena1108 the current code has been tested and works. I've tested the following actions:

* start a new service, check that the DNS entry is created
* scale the service up, check that the DNS entry is replaced without duplicated values
* scale the service down, check that old DNS entries are removed
* delete the service, check that the DNS entries are removed

As far as I'm concerned, it's functional. I'll start a PR for the catalog. I'm sure you'll still have comments about that code, so please let me know what I should optimize/fix.

File: providers/gandi.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	systemType := gandi.Production
+	testing := os.Getenv("GANDI_TESTING")
+	if len(testing) != 0 {
+		logrus.Infof("GANDI_TESTING is set, using testing platform")
+		systemType = gandi.Testing
+	}
+
+	if err := RegisterProvider("gandi", gandiHandler); err != nil {
```
Comment (alena1108) [inline providers/gandi.go]: looks like line 43-45 can be moved to the line 35. Sp if the provider fails to register, there is no need to process GANDI_TESTING env var

File: providers/gandi.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+	split_root := strings.Split(root, ".")
+	split_zoneDomain := split_root[len(split_root)-2:len(split_root)]
+	zoneDomain := strings.Join(split_zoneDomain, ".")
+	gandiHandler.zoneDomain = zoneDomain
+	gandiHandler.zoneSuffix = fmt.Sprintf(".%s", zoneDomain)
+	gandiHandler.sub = strings.TrimSuffix(root, zoneDomain)
+
+	domain := gandiDomain.New(client)
+	domainInfo, err := domain.Info(zoneDomain)
+	if err != nil {
```
Comment (alena1108) [inline providers/gandi.go]: Same for the blocks 61-63, 68-70 - would make sense to move them up. So if the error happens, we exit earlier before doing an unneeded settings on gandiHandler.

**Comment by raphink** [comment]
@alena1108 fixed both comments.

**Comment by alena1108** [comment]
@raphink LGTM. Could you update this PR with the link to the rancher-catalog PR, once its done? so I can merge both. The image tag would be v0.3.0

**Comment by raphink** [comment]
@alena1108 the rancher-catalog PR is done already, it's https://github.com/rancher/rancher-catalog/issues/90. I had already tagged it 0.3.0 since it was a new feature.

**Comment by alena1108** [comment]
@raphink awesome, thanks! I'll make a release/publish new template if not today, then early next week.

**Comment by raphink** [comment]
Great, thank you!

**Comment by alena1108** [comment]
@raphink could you squash all the commits into 1?

**Comment by raphink** [comment]
@alena1108 done

**Comment by alena1108** [comment]
@raphink just noticed that godeps weren't committed to https://github.com/rancher/external-dns/tree/master/Godeps/_workspace/src/github.com. Could you please add it? Modifying Godeps/Godeps.json is not enough

**Comment by raphink** [comment]
@alena1108 Just did that, but for some reason the gandi-api one doesn't seem to get added fully, because it uses submodules. How can I fix that?

File: Godeps/Godeps.json [hunk 3]
```diff
@@ -41,6 +45,23 @@
...
+		{
+			"ImportPath": "github.com/prasmussen/gandi-api/operation",
+			"Rev": "47911529ecd365be607afe09a4fe0799e16c3ed7"
+		},
+		{
+			"ImportPath": "github.com/prasmussen/gandi-api/util",
+			"Rev": "47911529ecd365be607afe09a4fe0799e16c3ed7"
+		},
+		{
+		{
```
Comment (alena1108) [inline Godeps/Godeps.json]: Seems like an error - extra {

Comment (raphink) [inline Godeps/Godeps.json]: Ah right, fixed.

**Comment by alena1108** [comment]
@raphink how have you added your godeps? Looks like prasmussen was added to godeps as a git submodule. it should have been done using "go get <projectname>" and then godep save. Try to re-add it by:

1) fix your godeps/godeps.json - there is one extra { there.
2) remove and re-add kolo and prasmussen dirs.
3) add the projects using:

go get github.com/prasmussen/gandi-api/util
...and other projects

4) run godep save ./...

Commit your changes and update the PR. You shouldn't see "Subproject commit 47911529ecd365be607afe09a4fe0799e16c3ed7" in 
Godeps/_workspace/src/github.com/prasmussen/gandi-api, but should see go files instead.

**Comment by alena1108** [comment]
Also remove these lines from your .git/config:

```
[branch "gandi"]
 15         remote = origin
 16         merge = refs/heads/gandi
```

**Comment by raphink** [comment]
I think everything is fine now, although when I run `godep restore`, I get an error about another module which I did not modify:

```
# cd /home/rpinson/go/src/rancher-external-dns/Godeps/_workspace/src/github.com/rancher/go-rancher; git checkout 465def4b0436df65693acbb3e3c6ed1bdec79e24
fatal: reference is not a tree: 465def4b0436df65693acbb3e3c6ed1bdec79e24
godep: error downloading dep (github.com/rancher/go-rancher/client): exit status 128
godep: Error downloading some deps. Aborting restore and check.
```

**Comment by raphink** [comment]
@alena1108 is that good for merging now?

**Comment by alena1108** [comment]
@raphink yup

**Pull request merged**